// @marinda/react-styled-buttons/dist/components/Button.stories.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.primary = exports["default"] = void 0;

var _core = require("@emotion/core");

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _addonA11y = require("@storybook/addon-a11y");

var _Button = _interopRequireDefault(require("./Button"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// variant knob
var variantName = "variant";
var variantValuesObj = {
  primary: "primary",
  secondary: "secondary",
  accent: "accent",
  accent2: "accent2",
  accent3: "accent3",
  accent4: "accent4",
  highlight: "highlight",
  muted: "muted",
  gray: "gray"
};
var variantOptionsObj = {
  display: "inline-radio"
};

var variantOptions = function variantOptions() {
  var variantDefaultValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "primary";
  return (0, _addonKnobs.optionsKnob)(variantName, variantValuesObj, variantDefaultValue, variantOptionsObj);
}; // size knob


var sizeLabel = "size";
var sizeValuesObj = {
  small: "small",
  medium: "medium",
  large: "large"
};
var sizeDefaultValue = "large";
var sizeOptionsObj = {
  display: "inline-radio"
};

var sizeOptions = function sizeOptions() {
  return (0, _addonKnobs.optionsKnob)(sizeLabel, sizeValuesObj, sizeDefaultValue, sizeOptionsObj);
}; // mode knob


var modeName = "mode";
var modeValuesObj = {
  light: "light",
  dark: "dark"
};
var modeDefaultValue = "light";
var modeOptionsObj = {
  display: "inline-radio"
};

var modeOptions = function modeOptions() {
  return (0, _addonKnobs.optionsKnob)(modeName, modeValuesObj, modeDefaultValue, modeOptionsObj);
}; // as knob


var asName = "as";
var asValuesObj = {
  button: "button",
  a: "a"
};
var asOptionsObj = {
  display: "inline-radio"
};

var asOptions = function asOptions() {
  var asDefaultValue = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "button";
  return (0, _addonKnobs.optionsKnob)(asName, asValuesObj, asDefaultValue, asOptionsObj);
};

var _default = {
  title: "Buttons|Button",
  decorators: [_addonKnobs.withKnobs, _addonA11y.withA11y]
};
exports["default"] = _default;

var primary = function primary() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
      _ref$children = _ref.children,
      children = _ref$children === void 0 ? "Primary Button" : _ref$children,
      onClick = _ref.onClick,
      disabled = _ref.disabled,
      _ref$disableColorTran = _ref.disableColorTransforms,
      disableColorTransforms = _ref$disableColorTran === void 0 ? false : _ref$disableColorTran,
      _ref$disableAnimation = _ref.disableAnimation,
      disableAnimation = _ref$disableAnimation === void 0 ? false : _ref$disableAnimation,
      _ref$id = _ref.id,
      id = _ref$id === void 0 ? "primary-button" : _ref$id,
      as = _ref.as;

  return (0, _core.jsx)(_Button["default"], {
    variant: variantOptions(),
    disabled: (0, _addonKnobs["boolean"])("Disabled", disabled),
    disableColorTransforms: (0, _addonKnobs["boolean"])("Disable color transforms", disableColorTransforms),
    disableAnimation: (0, _addonKnobs["boolean"])("Disable Animation", disableAnimation),
    size: sizeOptions(),
    mode: modeOptions(),
    as: asOptions(as),
    onClick: onClick,
    id: id
  }, children);
};

exports.primary = primary;
